src/edu/washington/escience/myria/operator/SamplingDistribution.java

```diff
@@ -0,0 +1,320 @@
+
+  /** Random generator used for creating the distribution. */
+  private Random rand;
+
+  /** Seed for the random generator. */
+  private Long randomSeed;
+
+  /**
+   * Distribution of the tuples across the workers. Value at index i == # of
+   * tuples on worker i.
+   */
+  ArrayList<Integer> tupleCounts;
+
+  /**
+   * Distribution of the actual stream size across the workers. May be different
+   * from tupleCounts if workers pre-sampled the data. Value at index i == # of
+   * tuples in stream on worker i.
+   */
+  ArrayList<Integer> streamCounts;
+
+  /** Total number of tuples across all workers. */
+  int totalTupleCount = 0;
+
+  private SamplingDistribution(Operator child, SamplingType sampleType,
+      Long randomSeed) {
+    super(child);
+    this.sampleType = sampleType;
+    this.randomSeed = randomSeed;
+  }
+
+  /**
+   * Instantiate a SamplingDistribution operator using a specific sample size.
+   *
+   * @param sampleSize
+   *          total samples to create a distribution for.
+   * @param sampleType
+   *          the type of sampling distribution to create
+   * @param child
+   *          extracts (WorkerID, PartitionSize, StreamSize) information from
+   *          this child.
+   * @param randomSeed
+   *          value to seed the random generator with. null if no specified seed
+   */
+  public SamplingDistribution(Operator child, int sampleSize,
+      SamplingType sampleType, Long randomSeed) {
+    this(child, sampleType, randomSeed);
+    this.sampleSize = sampleSize;
+    Preconditions.checkState(this.sampleSize >= 0,
+        "Sample Size must be >= 0: %s", this.sampleSize);
+  }
+
+  /**
+   * Instantiate a SamplingDistribution operator using a percentage of total
+   * tuples.
+   *
+   * @param samplePercentage
+   *          percentage of total samples to create a distribution for.
+   * @param sampleType
+   *          the type of sampling distribution to create
+   * @param child
+   *          extracts (WorkerID, PartitionSize, StreamSize) information from
+   *          this child.
+   * @param randomSeed
+   *          value to seed the random generator with. null if no specified seed
+   */
+  public SamplingDistribution(Operator child, float samplePercentage,
+      SamplingType sampleType, Long randomSeed) {
+    this(child, sampleType, randomSeed);
+    this.isPercentageSample = true;
+    this.samplePercentage = samplePercentage;
+    Preconditions.checkState(samplePercentage >= 0 && samplePercentage <= 100,
+        "Sample Percentage must be >= 0 && <= 100: %s", samplePercentage);
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws DbException {
+    // Drain out all the workerID and partitionSize info.
+    while (!getChild().eos()) {
+      TupleBatch tb = getChild().nextReady();
+      if (tb == null) {
+        if (getChild().eos()) {
+          break;
+        }
+        return null;
+      }
+      Type col0Type = tb.getSchema().getColumnType(0);
```

Comment:
I think you could factor out the sampling info extraction pretty cleanly into a separate helper (since I think it just mutates instance variables) and make this method more readable.